### ISSUE TITLE
Qubicle export crashes

### ISSUE THREAD

**Issue by Lynx3d** [guillaumechereau/goxel#85]
I just built Goxel from git (commit 6ce8d6d5e1099c5b848efe7f9f) because I'm looking for a nice voxel editor that export to qubicle, but it crashes every time I try to export something.

```
(gdb) bt
#0  0x000000000045b14e in mesh_iter (it=it@entry=0x7fffffffcd50, 
    pos=pos@entry=0x0) at src/mesh.c:619
#1  0x000000000051f675 in qubicle_export (mesh=0x6030007c3450, 
    path=path@entry=0x6040004bdbd0 "<removed>/untitled.qb") at src/formats/qubicle.c:136
#2  0x000000000051fd90 in export_as_qubicle (
    path=0x6040004bdbd0 "<removed>/untitled.qb")
    at src/formats/qubicle.c:179
#3  0x00000000004bdd79 in default_function (a=0x60d00000abe8, s=0x60d00003dd10)
    at src/action.c:73
#4  0x00000000004c1c79 in action_execv (action=action@entry=0x60d00000abe8, 
    sig=0x6c6381 "", sig@entry=0x6c6380 "", ap=ap@entry=0x7fffffffcfb0)
    at src/action.c:131
#5  0x00000000004c2303 in action_exec (action=0x60d00000abe8, sig=0x6c6380 "")
    at src/action.c:158
#6  0x000000000050ac94 in export_menu_action_callback (a=0x60d00000abe8, 
    user=0x0) at src/gui.cpp:1091
#7  0x00000000004c16e8 in actions_iter (
    f=0x50ab3b <export_menu_action_callback(action_t const*, void*)>, user=0x0)
    at src/action.c:49
#8  0x000000000050afed in render_menu () at src/gui.cpp:1122
#9  0x0000000000510258 in gui_iter (goxel=0x61a00001ec80, 
    inputs=0x7fffffffda30) at src/gui.cpp:1330

```

**Comment by guillaumechereau** [comment]
Thanks for pointing this out.  I just issued a fix.
Let me know if that doesn't work.

**Comment by Lynx3d** [comment]
Exporting succeeds now.
However trying to import those to Trove gives pretty scrambled results. I don't know yet which application is the issue here. Zoxel doesn't seem to be of any help either...

**Comment by guillaumechereau** [comment]
Is there a way to reproduce this problem without actually having a Trove account?

**Comment by Lynx3d** [comment]
Hm I think it should work, you can start Trove.exe directly (instead of launching from Glyph after logging in) and press F6 to enter the metaforge at the login prompt (which seems to be a non-functional leftover from before Glyph anyway).

The game comes with a "devtool_convert_to_blueprint.bat" that converts the .qb to .blueprint, which can than be loaded in the metaforge via drag&drop.

Though right now I suspect Trove doesn't really support qubicle files with multiple matrices, because I can't really see anything abviously wrong with the export code myself...

**Comment by guillaumechereau** [comment]
I see thanks.
I was actually thinking of changing the cubicle export to use a single matrix.  Maybe that would fix the bug?

**Comment by Lynx3d** [comment]
Just gave it a try to export in one piece, and it works. (Instead of rotating the whole thing I just swap x and z and save it right handed, seems to be prefered by zoxel now too, not sure why though)

`void qubicle_export(const mesh_t *mesh, const char *path)
{
    FILE *file;
    int x, y, z, vpos[3], size[3];
    uint8_t v[4];
    char buff[16];
    mesh_iterator_t iter;

    int bbox[2][3];
    mesh_get_bbox(mesh, bbox, true);

    file = fopen(path, "wb");
    WRITE(uint32_t, 257, file); // version
    WRITE(uint32_t, 0, file);   // color format RGBA
    WRITE(uint32_t, 1, file);   // orientation right handed
    WRITE(uint32_t, 0, file);   // no compression
    WRITE(uint32_t, 0, file);   // vmask
    WRITE(uint32_t, 1, file);

    sprintf(buff, "mtx1");
    WRITE(uint8_t, strlen(buff), file);
    fwrite(buff, strlen(buff), 1, file);
    iter =  mesh_get_accessor(mesh);
    size[0] = bbox[1][0] - bbox[0][0];
    size[1] = bbox[1][1] - bbox[0][1];
    size[2] = bbox[1][2] - bbox[0][2];
    WRITE(uint32_t, size[0], file);
    WRITE(uint32_t, size[2], file);
    WRITE(uint32_t, size[1], file);
    WRITE(int32_t, bbox[0][0], file);
    WRITE(int32_t, bbox[0][2], file);
    WRITE(int32_t, bbox[0][1], file);
    for (y = bbox[0][1]; y < bbox[1][1]; y++)
    for (z = bbox[0][2]; z < bbox[1][2]; z++)
    for (x = bbox[0][0]; x < bbox[1][0]; x++)
    {
        vpos[0] = x; vpos[1] = y; vpos[2] = z;
        mesh_get_at(mesh, &iter, vpos, v);
        fwrite(v, 4, 1, file);
    }
    fclose(file);
}`

**Comment by guillaumechereau** [comment]
OK thanks a lot!  I will change the code to work like that, it makes more sense! If you want to get proper attribution in git you can also make a pull request, up to you.

**Comment by Lynx3d** [comment]
Feel free to treat the code as your own, I just threw parts out and replaced them with bits and pieces I found to make it work without really understanding possible side effects.

Something that still confuses me is why the result would be right-handed, because from what I can tell, goxel already uses right-handed coordinates.

Also from what I found on the Qubicle website, a matrix is apparently some form of scene object that you define yourself to build compound objects, so maybe that is why everyone seems to work with just a single matrix for Trove. Furthermore there is a size limit of 1024x512x1024 in Qubicle, and they recommend to stay below 128³ for performance reasons, so possibly exporting as single matrix needs to be optional anyway...
